%\documentclass[3p,twocolumn]{article}
\documentclass[12pt]{article}

\usepackage{graphicx}
\usepackage{color}
\usepackage{url}
\usepackage{ifpdf}
\usepackage{hyperref}
\usepackage{xspace}
%\usepackage[draft]{pdfdraftcopy}

\setlength\parskip{-0.015em}
\setlength\parsep{-0.15em}

\newenvironment{shortlist}{
	\vspace*{-0.85em}
  \begin{itemize}
 \setlength{\itemsep}{-0.3em}
}{
  \end{itemize}
	\vspace*{-0.6em}
}

\usepackage{fullpage}
%\usepackage[top=tlength, bottom=blength, left=llength, right=rlength]{geometry} %http://en.wikibooks.org/wiki/LaTeX/Page_Layout
%\usepackage[margin=1in, paperwidth=5.5in, paperheight=8.5in]{geometry}

\usepackage{fancyhdr}
\setlength{\headheight}{16.0pt}
\pagestyle{fancy}
\headheight = 0pt
\headsep    = 25pt
\fancyhf{}
%\fancyhead[OC]{\bf {\it \footnotesize{Jha et al: A Case for SAGA as an Access Layer for DCI}}}

\newif\ifdraft
% \drafttrue
\ifdraft
 \newcommand{\amnote}[1]{  {\textcolor{magenta} {***AM: #1}}}
 \newcommand{\jhanote}[1]{ {\textcolor{red}     {***SJ: #1}}}
 \newcommand{\olenote}[1]{ {\textcolor{blue}    {***OW: #1}}}
\else
 \newcommand{\amnote}[1]{}
 \newcommand{\jhanote}[1]{}
 \newcommand{\olenote}[1]{}
\fi

\newcommand{\dn}{\vspace*{0.33em}}
\newcommand{\dnn}{\vspace*{0.66em}}
\newcommand{\dnnn}{\vspace*{1em}}
\newcommand{\uppp}{\vspace*{-1em}}
\newcommand{\upp}{\vspace*{-0.66em}}
\newcommand{\up}{\vspace*{-0.33em}}
\newcommand{\shift}{\hspace*{1.00em}}

\newcommand{\T}[1]{\texttt{#1}}
\newcommand{\I}[1]{\textit{#1}}
\newcommand{\B}[1]{\textbf{#1}}
\newcommand{\BI}[1]{\B{\I{#1}}}
\newcommand{\F}[1]{\B{[FIXME: #1]}}
\newcommand{\TODO}[1]{\textcolor{red}{\B{TODO: #1}}}

\begin{document}

\title{Towards a Scalable Architecture for Deep Sequencing Analytics}

\author{Joohyun Kim$^{1}$, Sharath Maddineni$^{1}$, Shantenu Jha$^{*1,2}$, \\
  \small{\emph{$^{1}$Center for Computation \& Technology, Louisiana State University, USA}}\\
  \small{\emph{$^{2}$Department of Computer Science, Louisiana State University, USA}}\\
  \small{\emph{$^{*}$Contact Author \texttt{sjha@cct.lsu.edu}}}
  }


\maketitle

\section*{Abstract}

We investigate the use of distributed computing environments, a
production HPC grid and a cloud environment for the genome-wide
mapping with BFast.  The main goal of this work is to understand
the characteristics of these two distributed computing environments
regarding computational requirements of the target bioinformatics
application whose computational complexity depends upon the size of a
reference genome, the data size of short reads from high throughput
technologies of the Next Generation Sequencing platforms, and their
biological genome contexts.

The DARE framework, comprising an open source Web application, Pylons
and the runtime environment for distributed scientific applications,
were employed for this work.  The framework enables us to develop a
lightweight, extensible, full-fledged science gateway effectively in
which the runtime environment built upon SAGA and BigJob abstraction
manages efficiently distributed computing-driven execution patterns of
target scientific applications.

Two model genomes, human genome and a microbe, Burkerholderia Glumae,
that might represent a eukaryote and a prokaryote system,
respectively, are chosen and two distributed environments, the
Louisiana Optical Network Initiative (LONI) grid and a Cloud system
from the FutureGrid, were used primarily focusing on different and
unique challenges in HPC and Cloud computing conditions.  Results are
analyzed by comparing advantages as well as limitations of each
distributed computing infrastructure in conjunction with distinctively
different two different genome systems.

Our results suggest the importance of an effective runtime environment
that facilitates a rapid development of cyberinfrastructure of
distributed computing resources and supports optimized execution
patterns for a target scientific application, in particular, of the
data-intensive genome-wide analysis.

\section{Introduction}

% \bibliographystyle{plain}
% \bibliography{egi-white-paper}

\end{document}

